docs/SIL.rst

```diff
@@ -1713,9 +1713,96 @@ normal uses always before consuming uses. Any such violations of ownership
+                +----+
+      +-------- |None| -----------+
+      |         +----+            |
+      |            |              |
+      v            v              v         ^
+  +-------+  +-----+------+  +---------+    |
+  | Owned |  | Guaranteed |  | Unowned |    +--- Value Ownership Kinds and
+  +-------+  +-----+------+  +---------+         Ownership Constraints
+      |            |              |
+      |            v              |         +--- Only Ownership Constraints
+      |          +---+            |         |
+      +--------->|Any|<-----------+         v
+                 +---+
+
+One moves down the lattice by performing a "meet" operation::
+
+  None meet OtherOwnershipKind -> OtherOwnershipKind
+  Unowned meet Owned -> Any
+  Owned meet Guaranteed -> Any
+
+and one moves up the lattice by performing a "join" operation, e.x.::
+
+  Any join OtherOwnershipKind -> OtherOwnershipKind
+  Owned join Any -> Owned
+  Owned join Guaranteed -> None
+
+This lattice is applied to SIL by requiring well formed SIL to:
+
+1. Define a static map of each SIL value to an OwnershipKind that classify the
+   semantics that the SIL value obeys. We call this subset of OwnershipKind to
+   be the set of `Value Ownership Kind`_: `None`_, `Unowned`_, `Guaranteed`_,
+   `Owned`_ (note conspiciously missing `Any`_). This is because in our model
+   `Any`_ represents an unknown ownership semantics and since our model is
+   statically strict, we do not allow for values to have unknown ownership.
+
+2. Define a static map from each operand of a SILInstruction, `i`, to an
+   (Ownership Kind, Boolean) called the operand's `Ownership Constraint`_. The
+   Ownership Kind element of the `Ownership Constraint`_ acts as a a static
+   "constraint" on the Value Ownership Kind that the operand's value can be mapped
+   to. The second boolean value is used to know if an operand will end the
+   lifetime of the incoming value when checking dataflow rules. The dataflow
+   rules that each `Value Ownership Kind`_ obeys is documented for each
+   `Value Ownership Kind`_ in its detailed description below.
+
+Then we take these static definitions and require that valid SIL has the
+property that given a value ``v``, an instruction ``i`` and said instruction's
+operand ``operand(i)``, that ``v`` can be used as a value in ``operand(i)`` only
+if the join of the ownership constraint of ``operand(i)`` with ``v`` is the
+ownership kind of ``v``! In symbols, we must have the following join is
+defined::
+
+  join : (OwnershipConstraint, ValueOwnershipKind) -> ValueOwnershipKind
+  OwnershipConstraint(operand(i)) join ValueOwnershipKind(v) = ValueOwnershipKind(v)
+
+Operationally this has a few interesting effects on SIL::
+
+1. We have defined away invalid value-operand (aka def-use) pairing since
+   ``OwnershipConstraint(operand) meet ValueOwnershipKind(v)`` would be ``Any``
+   which would be invalid for a value to take on, implying that the SIL would be
+   invalid.
```

Comment:
To be consistent with the previous paragraph, I think this should just say that the `join` property stated above ensures no invalid def-use pairing. Talking about performing a `meet` on the constraint actually confuses me.

However, it is interesting to say *additionally* that this formulation of the problem defines away combining incompatible ownership via aggregates or phi values because the `meet` over those values will result in `Any` ownership kind, which is an invalid value ownership.

Reply:
Yes. But then I would need to talk about Ownership Phis.

Reply:
Maybe I should add a section on Ownership Phis next.

Reply:
You could introduce `meet` as being used when combining multiple owned values here or not. It just doesn't make sense to me to say `OwnershipConstraint meet ValueOwnership` For example, it's invalid to pass an owned value into a `None` constraint, but the meet won't give you `Any`

Reply:
Ok. I removed the meet comment and replaced it with join. I really meant to talk about join here.